### PR TITLE
Support padding during allocation

### PR DESCRIPTION
**Pull request by deyuan**
Partial implementation:
- Allow bitsPerElement to be smaller than data type bit width
- Allow a narrower data type to be associated to a wider data type

Need to revisit all code depending on bitsPerElement to make it work.

Related to Issue #16

### CODE REVIEW

File: libpimeval/src/pimResMgr.cpp [hunk 1]
```diff
@@ -102,6 +102,9 @@ pimResMgr::~pimResMgr()
 }
 
 //! @brief  Alloc a PIM object
+//!         Expect bitsPerElement to be smaller than or equal to the width of dataType
+//!         For V layout, allocate bitsPerElement rows
+//!         For H layout, pad element to the width of dataType
 PimObjId
 pimResMgr::pimAlloc(PimAllocEnum allocType, uint64_t numElements, unsigned bitsPerElement, PimDataType dataType)
```
Comment (fasiddique) [inline libpimeval/src/pimResMgr.cpp:109]: Do we actually need `bitsPerElement`? Isn't it redundant? We can derive bits per element from the datatype.

File: libpimeval/src/pimResMgr.cpp [hunk 2]
```diff
@@ -102,6 +102,9 @@ pimResMgr::~pimResMgr()
 }
 
 //! @brief  Alloc a PIM object
+//!         Expect bitsPerElement to be smaller than or equal to the width of dataType
```
Comment (fasiddique) [inline libpimeval/src/pimResMgr.cpp:105]: I am wondering why this should be smaller that the datatype width?

File: libpimeval/src/pimResMgr.cpp (hunk 1 above)
Comment (deyuan) [reply libpimeval/src/pimResMgr.cpp:109]: This is especially needed for V layout architectures. We support a set of common data types, but an application may just use fewer bits of it. For example, we can allocate an INT32 object but only use 20 bits of it. In this case, this bitsPerElement is essential for performance modeling.

File: libpimeval/src/pimResMgr.cpp (hunk 2 above)
Comment (deyuan) [reply libpimeval/src/pimResMgr.cpp:105]: We need to support both V and H layouts. Data types determines the allocated bit width, while bitsPerElements determines how many bits are actually used.

Comment (fasiddique) [reply libpimeval/src/pimResMgr.cpp:105]: This approach seems quite confusing and increases the risk of introducing bugs. If someone is creating a vector of int32, they should expect it to use 32 bits. If a smaller bit-size is needed, they should opt for a different data type that uses fewer than 32 bits.

File: libpimeval/src/pimResMgr.cpp (hunk 1 above)
Comment (fasiddique) [reply libpimeval/src/pimResMgr.cpp:109]: Then we should introduce INT20. I don't think this is a good design at all.

Comment (deyuan) [reply libpimeval/src/pimResMgr.cpp:109]: I think we've discussed the following in emails. We need to support pimAllocAssociate with different data types, and PIM operations with mixed data types. For example, we may allocate an INT32 object first, then associate an INT8 or BOOL object with it. In this case, we would like to do element padding and align elements with different types. Each element will have a data width and a padded width. The code changes here are for supporting such functionality. We can remove bitsPerElements from the functional API later for simplicity, but I don't think code design is bad here

File: libpimeval/src/pimResMgr.cpp (hunk 2 above)
Comment (deyuan) [reply libpimeval/src/pimResMgr.cpp:105]: Let's discuss how to support element padding when you are back

**Comment by deyuan** [comment]
Switch development to dev-alloc branch

**Pull request closed without merging**